## Pandas. Практика

In [ ]:
import numpy as np
import pandas as pd

Описание столбцов датасета:
- `date` – дата, когда был совершен заказ;
- `vendor_id` – уникальный идентификатор ресторана;
- `chain_id` – уникальный идентификатор сети ресторанов;
- `city_id` – уникальный идентификатор города;
- `spec` – специализация ресторана;
- `successful_orders` – число успешных заказов;
- `fail_orders` – число неуспешных заказов.

In [ ]:
df = pd.read_csv('orders.csv.gz', parse_dates=['date'], compression='gzip')
df.head(5)

date  vendor_id  chain_id  city_id     spec  successful_orders   
0 2019-07-02      40065      7501       23     Суши               54.0  \
1 2019-08-27      48058     33226       23   Шаурма                0.0   
2 2019-09-25      35430     26220       25    Пицца                2.0   
3 2019-09-21      56553     38601       23  Шашлыки                1.0   
4 2019-09-21      43919     30984       25     Суши                6.0   

   fail_orders  
0          1.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0

### Задание 1 (0.5 балла)

Выведите информацию о том, сколько оперативной памяти занимает загруженный датасет.

In [ ]:
# место для вашего кода

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96118 entries, 0 to 96117
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   date               96118 non-null  datetime64[ns]
 1   vendor_id          96118 non-null  int64         
 2   chain_id           96118 non-null  int64         
 3   city_id            96118 non-null  int64         
 4   spec               95733 non-null  object        
 5   successful_orders  96118 non-null  float64       
 6   fail_orders        96118 non-null  float64       
dtypes: datetime64[ns](1), float64(2), int64(3), object(1)
memory usage: 5.1+ MB


### Задание 2 (0.5 балла)

Выведите описательные статистики для датасета.

In [26]:
df.describe(include='all')

date     vendor_id      chain_id  \
count                           96118  96118.000000  96118.000000   
unique                            NaN           NaN           NaN   
top                               NaN           NaN           NaN   
freq                              NaN           NaN           NaN   
mean    2019-08-02 22:08:09.748017920  47425.345867  31031.897074   
min               2019-06-02 00:00:00   3756.000000    804.000000   
25%               2019-07-04 00:00:00  40265.000000  27909.000000   
50%               2019-08-03 00:00:00  49754.000000  32662.000000   
75%               2019-09-02 00:00:00  58052.000000  38376.000000   
max               2019-09-30 00:00:00  76557.000000  47907.000000   
std                               NaN  15447.399597  10669.949918   

             city_id   spec  successful_orders   fail_orders  
count   96118.000000  95733       96118.000000  96118.000000  
unique           NaN     35                NaN           NaN  
top              NaN   Суши                NaN           NaN  
freq             NaN  23077                NaN           NaN  
mean       24.357342    NaN           8.877401      0.416842  
min        23.000000    NaN           0.000000      0.000000  
25%        23.000000    NaN           1.000000      0.000000  
50%        24.000000    NaN           4.000000      0.000000  
75%        25.000000    NaN           9.000000      0.000000  
max        26.000000    NaN         240.000000     86.000000  
std         1.168019    NaN          15.370105      1.424931

### Задание 3 (1 балл)

Сколько уникальных специализаций присутствует в нашем датасете? Дополнительно выведите их список и частоту встречаемости.

In [38]:
# Количество уникальных специализаций
print(df['spec'].nunique())

35


In [39]:
# Список и частота
df['spec'].value_counts()

spec
Суши                   23077
Пицца                  13825
Бургеры                13555
Русская                 6260
Шашлыки                 5658
Шаурма                  5230
Мясо                    3876
Фастфуд                 3813
Сэндвичи                3375
Грузинская              3043
Итальянская             2399
Японская                1577
Выпечка                 1205
Рыба                    1103
Десерты                 1071
Китайская                841
Узбекская                790
Обеды                    786
Паназиатская             528
Пироги                   467
Здоровая еда             456
Кофе и чай               453
Супы                     405
Паста и ризотто          387
Вьетнамская              382
Завтраки                 346
Корейская                202
Индийская                194
Хот-доги                 120
Гирос                    108
Мексиканская             103
Хумус                     34
Вегетарианское меню       28
Халяль                    28
Хинкали  

### Задание 4 (1 балл)

Приведите столбец специализации к типу категории. Выведите новые типы для всех столбцов в датасете.

In [ ]:
# место для вашего кода

date                 datetime64[ns]
vendor_id                     int64
chain_id                      int64
city_id                       int64
spec                       category
successful_orders           float64
fail_orders                 float64
dtype: object

### Задание 5 (1 балл)

Посчитайте, сколько было дней у ресторана с номером `40065`, когда количество успешных заказов было менее 20.

In [ ]:
# место для вашего кода

45

### Задание 6 (1 балл)

Посчитайте среднее количество успешно выполненных заказов рестораном в сутки.

In [ ]:
# место для вашего кода

8.878465140478667

### Задание 7 (1 балл)

Выведите города **по убыванию** количества успешных заказов за все время.

In [ ]:
# место для вашего кода

city_id
23    277744.0
26    232039.0
25    228518.0
24    114977.0
Name: successful_orders, dtype: float64

### Задание 8 (2 балла)

Ответьте на вопрос:

> Какие специализации имеют самую высокую долю неуспешных заказов за весь период времени?

Доля неуспешных заказов рассчитывается как отношение неуспешных заказов к общему числу заказов (успешные + неуспешные). Выведите топ-5 таких спеицализаций.

In [ ]:
# место для вашего кода

fail_orders  total_orders  ptg_fail_orders
spec                                                           
Вегетарианское меню         24.0          29.0         0.827586
Халяль                      10.0          46.0         0.217391
Хинкали                      2.0          12.0         0.166667
Завтраки                    88.0         566.0         0.155477
Хот-доги                    57.0         405.0         0.140741

### Задание 9 (2 балла)

Ответьте на вопрос:

> Какая из сетей ресторанов (`chain_id`) показала максимальный абсолютный рост за август 2019 года по сравнению с июлем 2019 года?

При ответе на вопрос стоит учитывать только сети, у которых заказы были в августе (заказов в июле могло не быть). Для решения этой задачи используйте join-ы.

In [ ]:
# место для вашего кода

chain_id  successful_orders_aug  successful_orders_jul  increase
207     29925                12561.0                11167.0    1394.0